# Projeto 1 - Ciência dos Dados

Nome: __Diego___

Nome: _Joseph____

Nome: _Guido____

Atenção: Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [1]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

In [2]:
print('Esperamos trabalhar no diretório')
print(os.getcwd())

Esperamos trabalhar no diretório
C:\Users\Diego


Carregando a base de dados com os tweets classificados como relevantes e não relevantes:

In [3]:
filename = 'DONDA.xlsx'

In [4]:
train = pd.read_excel(filename)
train.head(5)

Treinamento  CLASSIFICACAO  \
0  se os gringos tem kenye west com donda\nos br ...              5   
1              @deloelmo era menos negrero que donda              4   
2                             indo escutar donda abs              3   
3                        @rapmais creditado em donda              3   
4                     donda aoty mesmo nao tem jeito              4   

   Unnamed: 2  Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN   

        Unnamed: 7 Unnamed: 8  
0       CATEGORIAS        NUM  
1  MUITO RELEVANTE          5  
2        RELEVANTE          4  
3           NEUTRO          3  
4      IRRELEVANTE          2

In [5]:
test = pd.read_excel(filename, sheet_name = 'Teste')
test.head(5)

Teste  CLASSIFICACAO  \
0  08:21, tomando meu café, ouvindo donda e escre...              2   
1                       @ali3nobserver nossa f donda              4   
2  pra mim o top 3 eh\ncall me if you get lost\nc...              5   
3  @wowhespittin não é para compararem donda com ...              5   
4  tem duas semanas que só escuto o donda com uma...              5   

   CLASSIFICACAO_MACHINE  Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0                    NaN         NaN         NaN         NaN         NaN   
1                    NaN         NaN         NaN         NaN         NaN   
2                    NaN         NaN         NaN         NaN         NaN   
3                    NaN         NaN         NaN         NaN         NaN   
4                    NaN         NaN         NaN         NaN         NaN   

   Unnamed: 7         CATEGORIAS  NUM  
0         NaN    MUITO RELEVANTE  5.0  
1         NaN          RELEVANTE  4.0  
2         NaN             NEUTRO  3.0  
3         NaN        IRRELEVANTE  2.0  
4         NaN  MUITO IRRELEVANTE  1.0

___
## Classificador automático de sentimento


Faça aqui uma descrição do seu produto e o que considerou como relevante ou não relevante na classificação dos tweets.

ESCREVA AQUI...

O produto escolhido foi o novo albúm do artista Kanye West, um dos maiores do mundo, chamado Donda. O albúm era muito esperado pelos fãs devido a demora do lançamento, então queríamos analisar como eles reagiram às novas músicas. Além disso, queríamos verificar se houveram comparações entre esse albúm e o novo albúm do artista Drake, lançado em uma data bem próxima ao de Kanye. Para isso, determinamos 5 categorias: Muito Relevante, Relevante, Neutro, Irrelevante e Muito Irrelevante. Nas duas primeiras categorias, separamos os tweets que expressavam a opinião das pessoas sobre o albúm, contendo palavras como: 'gostei, amei, não paro de escutar' etc, e também tweets que continham tanto 'Donda' como 'CLB'(novo albúm do Drake). Na categoria neutro elencamos tweets que não expressavam uma opinião muito explícita sobre nenhum dos dois albúns, ou apenas mencionavam Donda. Nas duas últimas categorias, colocamos os tweets que não apenas mencionavam o novo sem contexto algum, ou até sem estar falando do próprio albúm.

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [6]:
all_words = []
all_wordsN = []
all_words_5 = []
all_words_4 = []
all_words_3 = []
all_words_2 = []
all_words_1 = []

#CLEAN UP

import re 
def cleanup(train):
    punctuation = '[!-.:?;''()]' 
    pattern = re.compile(punctuation)
    text_subbed = re.sub(pattern, '', train)
    return text_subbed

#REMOVER REPETIDOS PARA O ALL_WORDS

def remove_repetidos(lista):
    l = []
    for i in lista:
        if i not in l:
            l.append(i)
    l.sort()
    return l

#LISTA TODAS PALAVRAS

provisoria = []
for tweet in train['Treinamento']:
    splitado = tweet.split()
    for palavra in splitado:
        palavra_clean = cleanup(palavra)
        provisoria.append(palavra_clean)
all_words = remove_repetidos(provisoria) #all_words eh uma lista sem palavras repetidas agora

#NOTA 5
        
rate_5 = train['CLASSIFICACAO'] == 5
tweets_5 = train.loc[rate_5, :]
for tweet in tweets_5['Treinamento']:
    splitado = tweet.split()
    for palavra in splitado:
        palavra_clean = cleanup(palavra)
        all_words_5.append(palavra_clean)

#NOTA 4
       
rate_4 = train['CLASSIFICACAO'] == 4
tweets_4 = train.loc[rate_4, :]
for tweet in tweets_4['Treinamento']:
    splitado = tweet.split()
    for palavra in splitado:
        palavra_clean = cleanup(palavra)
        all_words_4.append(palavra_clean)

#NOTA 3
       
rate_3 = train['CLASSIFICACAO'] == 3
tweets_3 = train.loc[rate_3, :]
for tweet in tweets_3['Treinamento']:
    splitado = tweet.split()
    for palavra in splitado:
        palavra_clean = cleanup(palavra)
        all_words_3.append(palavra_clean)

#NOTA 2
       
rate_2 = train['CLASSIFICACAO'] == 2
tweets_2 = train.loc[rate_2, :]
for tweet in tweets_2['Treinamento']:
    splitado = tweet.split()
    for palavra in splitado:
        palavra_clean = cleanup(palavra)
        all_words_2.append(palavra_clean)

#NOTA 1
       
rate_1 = train['CLASSIFICACAO'] == 1
tweets_1 = train.loc[rate_1, :]
for tweet in tweets_1['Treinamento']:
    splitado = tweet.split()
    for palavra in splitado:
        palavra_clean = cleanup(palavra)
        all_words_1.append(palavra_clean)

In [7]:
def nota(tweet):
    tweet_words_uncleaned = tweet.split()
    tweet_words = [] #palavras limpas do tweet
    for palavra in tweet_words_uncleaned:
        tweet_words.append(cleanup(palavra))
    P5 = 1
    P4 = 1
    P3 = 1
    P2 = 1
    P1 = 1
    for word in tweet_words:
        frequencia_em_5 = all_words_5.count(word)
        P5_word = (frequencia_em_5+1)/(len(all_words_5)+len(all_words))
        P5 *= P5_word
        frequencia_em_4 = all_words_4.count(word)
        P4_word = (frequencia_em_4+1)/(len(all_words_4)+len(all_words))
        P4 *= P4_word
        frequencia_em_3 = all_words_3.count(word)
        P3_word = (frequencia_em_3+1)/(len(all_words_3)+len(all_words))
        P3 *= P3_word
        frequencia_em_2 = all_words_2.count(word)
        P2_word = (frequencia_em_2+1)/(len(all_words_2)+len(all_words))
        P2 *= P2_word
        frequencia_em_1 = all_words_1.count(word)
        P1_word = (frequencia_em_1+1)/(len(all_words_1)+len(all_words))
        P1 *= P1_word
    lista_comparacao = [P1, P2, P3, P4, P5]
    maior = max(lista_comparacao)
    if maior == P5:
        return 5
    elif maior == P4:
        return 4
    elif maior == P3:
        return 3
    elif maior == P2:
        return 2
    elif maior == P1:
        return 1

In [8]:
j = 0
for tweet in test['Teste']:
    rate = nota(tweet)
    test['CLASSIFICACAO_MACHINE'][j] = rate
    j+=1
test.head(15)

<ipython-input-8-976a172372ed>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test['CLASSIFICACAO_MACHINE'][j] = rate


Teste  CLASSIFICACAO  \
0   08:21, tomando meu café, ouvindo donda e escre...              2   
1                        @ali3nobserver nossa f donda              4   
2   pra mim o top 3 eh\ncall me if you get lost\nc...              5   
3   @wowhespittin não é para compararem donda com ...              5   
4   tem duas semanas que só escuto o donda com uma...              5   
5    fr donda esta paramore!? https://t.co/2t2n8chfll              2   
6   só aceitaria tyler perdendo se fosse contra do...              4   
7   castelos e ruínas &gt;&gt;&gt;&gt;&gt;&gt;&gt;...              3   
8   @afonso_palhinha e o clb n conta tal como o donda              4   
9   acabou a luz aqui ficou tudo igual a capa do d...              1   
10  ah pq donda kanye west o album novo do drake o...              4   
11         até hoje não criei coragem pra ouvir donda              4   
12  só tsu do drake bota o donda no bolso  falei t...              4   
13  donda não precisava de nenhuma música parte do...              5   
14  porque enalteceram tanto "donda" e ninguém fal...              5   

    CLASSIFICACAO_MACHINE  Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0                     4.0         NaN         NaN         NaN         NaN   
1                     1.0         NaN         NaN         NaN         NaN   
2                     5.0         NaN         NaN         NaN         NaN   
3                     4.0         NaN         NaN         NaN         NaN   
4                     5.0         NaN         NaN         NaN         NaN   
5                     1.0         NaN         NaN         NaN         NaN   
6                     5.0         NaN         NaN         NaN         NaN   
7                     3.0         NaN         NaN         NaN         NaN   
8                     4.0         NaN         NaN         NaN         NaN   
9                     1.0         NaN         NaN         NaN         NaN   
10                    5.0         NaN         NaN         NaN         NaN   
11                    4.0         NaN         NaN         NaN         NaN   
12                    3.0         NaN         NaN         NaN         NaN   
13                    2.0         NaN         NaN         NaN         NaN   
14                    5.0         NaN         NaN         NaN         NaN   

    Unnamed: 7         CATEGORIAS  NUM  
0          NaN    MUITO RELEVANTE  5.0  
1          NaN          RELEVANTE  4.0  
2          NaN             NEUTRO  3.0  
3          NaN        IRRELEVANTE  2.0  
4          NaN  MUITO IRRELEVANTE  1.0  
5          NaN                NaN  NaN  
6          NaN                NaN  NaN  
7          NaN                NaN  NaN  
8          NaN                NaN  NaN  
9          NaN                NaN  NaN  
10         NaN                NaN  NaN  
11         NaN                NaN  NaN  
12         NaN                NaN  NaN  
13         NaN                NaN  NaN  
14         NaN                NaN  NaN

___
### Verificando a performance do Classificador

Agora você deve testar o seu classificador com a base de Testes.

In [9]:
Acuracia = 0
Diferentes_maiores = 0
Diferentes_menor = 0
Total = 0
i = 0

for tweet in test['Teste']:
    if (test['CLASSIFICACAO_MACHINE'][i]) > (test['CLASSIFICACAO'][i]):
        Diferentes_maiores+=1
        Total +=1
        i += 1        
    elif (test['CLASSIFICACAO_MACHINE'][i]) < (test['CLASSIFICACAO'][i]):
        Diferentes_menor+=1
        Total +=1
        i += 1
    elif (test['CLASSIFICACAO_MACHINE'][i]) == (test['CLASSIFICACAO'][i]):
        Acuracia +=1
        Total +=1
        i += 1
        
P_acuracia = (Acuracia/Total)*100
P_Diferentes_maiores = (Diferentes_maiores/Total)*100
P_Diferentes_menores = (Diferentes_menor/Total)*100
print('A Acurácia do Classificador foi de {0}%, o classificador elencou {1}% dos tweets com relevância maior que nós e {2}% com relevância menor'.format(P_acuracia,P_Diferentes_maiores,P_Diferentes_menores))

    

A Acurácia do Classificador foi de 36.0%, o classificador elencou 37.5% dos tweets com relevância maior que nós e 26.5% com relevância menor


In [10]:
MI_I = 0
MI_N = 0
MI_R = 0
MI_MR = 0

i=0
for tweet in test['Teste']:
    if (test['CLASSIFICACAO'][i]) == 1:
        if (test['CLASSIFICACAO_MACHINE'][i]) == 2:
            MI_I += 1
            i+=1
        elif (test['CLASSIFICACAO_MACHINE'][i]) == 1:
            i+=1
        elif (test['CLASSIFICACAO_MACHINE'][i]) == 3:
            MI_N += 1
            i+=1
        elif (test['CLASSIFICACAO_MACHINE'][i]) == 4:
            MI_R += 1
            i+=1
        elif (test['CLASSIFICACAO_MACHINE'][i]) == 5:
            MI_MR += 1
            i+=1
    else:   
        i+=1
        
Total1 = MI_I + MI_N + MI_R + MI_MR

print('Erros do Muito Irrelevante: \n {0} MI-> I,\n {1} MI--> N,\n {2} MI--> R,\n {3} MI --> MR, \n Total: {4}'.format(MI_I, MI_N, MI_R, MI_MR,Total1))

Erros do Muito Irrelevante: 
 2 MI-> I,
 3 MI--> N,
 4 MI--> R,
 1 MI --> MR, 
 Total: 10


In [11]:
I_MI = 0
I_N = 0
I_R = 0
I_MR = 0
j = 0
for tweet in test['Teste']:
    if (test['CLASSIFICACAO'][j]) == 2:
        if (test['CLASSIFICACAO_MACHINE'][j]) == 1:
            I_MI += 1
            j+=1
        elif (test['CLASSIFICACAO_MACHINE'][j]) == 3:
            I_N += 1
            j+=1
        elif (test['CLASSIFICACAO_MACHINE'][j]) == 2:
            j+=1
        elif (test['CLASSIFICACAO_MACHINE'][j]) == 4:
            I_R += 1
            j+=1
        elif (test['CLASSIFICACAO_MACHINE'][j]) == 5:
            I_MR += 1
            j+=1
    else:
        j+=1
Total2 = I_MI + I_N + I_R + I_MR
print('Erros do Irrelevante: \n {0} I --> MR,\n {1} I--> N,\n {2} I--> R,\n {3} I --> MR, \n Total: {4}'.format(I_MI, I_N, I_R, I_MR,Total2))

Erros do Irrelevante: 
 6 I --> MR,
 7 I--> N,
 16 I--> R,
 4 I --> MR, 
 Total: 33


In [12]:
N_MI = 0
N_I = 0
N_R = 0
N_MR = 0
m = 0
for tweet in test['Teste']:
    if (test['CLASSIFICACAO'][m]) == 3:
        if (test['CLASSIFICACAO_MACHINE'][m]) == 1:
            N_MI += 1
            m+=1
        elif (test['CLASSIFICACAO_MACHINE'][m]) == 2:
            N_I += 1
            m+=1
        elif (test['CLASSIFICACAO_MACHINE'][m]) == 3:
            m+=1
        elif (test['CLASSIFICACAO_MACHINE'][m]) == 4:
            N_R += 1
            m+=1
        elif (test['CLASSIFICACAO_MACHINE'][m]) == 5:
            N_MR += 1
            m+=1
    else:
        m +=1
Total3 = N_MI + N_I +N_R + N_MR
print('Erros do Neutro: \n {0} N --> MR,\n {1} N--> I,\n {2} N--> R,\n {3} N --> MR, \n Total: {4}'.format(N_MI, N_I, N_R, N_MR, Total3))

Erros do Neutro: 
 3 N --> MR,
 1 N--> I,
 14 N--> R,
 4 N --> MR, 
 Total: 22


In [13]:
R_MI = 0
R_I = 0
R_N = 0
R_MR = 0
n = 0
for tweet in test['Teste']:
    if (test['CLASSIFICACAO'][n]) == 4:
        if (test['CLASSIFICACAO_MACHINE'][n]) == 1:
            R_MI += 1
            n+=1
        elif (test['CLASSIFICACAO_MACHINE'][n]) == 2:
            R_I += 1
            n+=1
        elif (test['CLASSIFICACAO_MACHINE'][n]) == 3:
            R_N += 1
            n+=1
        elif (test['CLASSIFICACAO_MACHINE'][n]) == 4:
            n+=1
        elif (test['CLASSIFICACAO_MACHINE'][n]) == 5:
            R_MR += 1
            n+=1
    else:
        n+=1
        
Total4 = R_MI + R_I +R_N + R_MR
print('Erros do Relevante: \n {0} R --> MI,\n {1} R--> I,\n {2} R--> N,\n {3} R --> MR, \n Total: {4}'.format(R_MI, R_I, R_N, R_MR,Total4))

Erros do Relevante: 
 5 R --> MI,
 2 R--> I,
 11 R--> N,
 20 R --> MR, 
 Total: 38


In [14]:
MR_MI = 0
MR_I = 0
MR_N = 0
MR_R = 0
p = 0
for tweet in test['Teste']:
    if (test['CLASSIFICACAO'][p]) == 5:
        if (test['CLASSIFICACAO_MACHINE'][p]) == 1:
            MR_MI += 1
            p+=1
        elif (test['CLASSIFICACAO_MACHINE'][p]) == 2:
            MR_I += 1
            p+=1
        elif (test['CLASSIFICACAO_MACHINE'][p]) == 3:
            MR_N += 1
            p+=1
        elif (test['CLASSIFICACAO_MACHINE'][p]) == 5:
            p+=1
        elif (test['CLASSIFICACAO_MACHINE'][p]) == 4:
            MR_R += 1
            p+=1
    else:
        p+=1
        
Total5 = MR_MI + MR_I +MR_N + MR_R
print('Erros do Muito Relevante: \n {0} MR --> MI,\n {1} MR--> I,\n {2} MR--> N,\n {3} MR --> R, \n Total: {4}'.format(MR_MI, MR_I, MR_N, MR_R,Total5))

Erros do Muito Relevante: 
 0 MR --> MI,
 2 MR--> I,
 3 MR--> N,
 20 MR --> R, 
 Total: 25


___
### Concluindo

Testando o nosso classificador que tinha como principal objetivo de separar os tweets contidos em uma base de dados que expressavam a opinião das pessoas sobre o albúm Donda de Kanye West, contendo palavras como: 'gostei, amei, não paro de escutar' etc, e também tweets que continham tanto 'Donda' como 'CLB'(novo albúm do Drake), verificado a comparação que as pessoas faziam entre os dois. 

Ao testar o classificador considerando apenas as mensagens da planilha Teste, o objetivo é verificar se o classificador entendeu quais são as palavras mais comuns (frequentes) em uma mensagem muito relevante, as mais presentes nas mensagens relevantes e assim por diante.

Testando o classificador verificamos uma acurácia de 36%, ou seja o classificador classificou 36% dos tweets da mesma forma que nós fizemos manualmente na planilha do excel. Dos que ele classificou diferente foram 37.5% dos tweets com relevância maior que nós e 26.5% com relevância menor. Depois analisando cada categoria por si só, notamos poucos erros na classificação dos tweets Muito Irrelevantes, mostrando eficácia do classificador em descartar tweets que não agregam ao objeto de análise. Quando analisando os tweets Neutros, vimos que a maioria dos erros do classificador foi em classificar os Nuetros em Relevantes, portanto um erro plausível. O mesmo acontece na classificação dos tweets Relevantes e Muito relevantes, em que a maioria dos erros do nosso classificador foi em classificá-los em Muito Relevantes e Relevantes respectivamente. A maior ineficiencia do classificador foi em classificar os tweets Irrelevante, já que classificou uma grande parcela dos tweets em Relevantes e Muito Relevantes. Vendo esses erros, fomos verificar a planilha e consiguimos perceber que uma parte dos erros não foi do classificador, mas nosso ao elencar manualmente a planilha, de forma que o classificador estava certo em grande parte dos casos. 
Levando tudo isso em consideração podemos considerar nosso classificador eficaz.

___
### Qualidade do Classificador a partir de novas separações dos tweets entre Treinamento e Teste

Caso for fazer esse item do Projeto

___
## Aperfeiçoamento:

Trabalhos que conseguirem pelo menos conceito B vão evoluir em conceito dependendo da quantidade de itens avançados:

* IMPLEMENTOU outras limpezas e transformações que não afetem a qualidade da informação contida nos tweets. Ex: stemming, lemmatization, stopwords
* CORRIGIU separação de espaços entre palavras e emojis ou entre emojis e emojis
* CRIOU categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante. Pelo menos quatro categorias, com adição de mais tweets na base, conforme enunciado. (OBRIGATÓRIO PARA TRIOS, sem contar como item avançado)
* EXPLICOU porquê não pode usar o próprio classificador para gerar mais amostras de treinamento
* PROPÔS diferentes cenários para Naïve Bayes fora do contexto do projeto
* SUGERIU e EXPLICOU melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)
* FEZ o item 6. Qualidade do Classificador a partir de novas separações dos tweets entre Treinamento e Teste descrito no enunciado do projeto (OBRIGATÓRIO para conceitos A ou A+)

___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**